filters/ReturnsFilter.cpp

```diff
@@ -0,0 +1,126 @@
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+ * OF SUCH DAMAGE.
+ ****************************************************************************/
+
+#include "ReturnsFilter.hpp"
+
+#include <pdal/util/ProgramArgs.hpp>
+
+namespace pdal
+{
+
+static StaticPluginInfo const s_info{
+    "filters.returns", "Split data by return order",
+    "http://pdal.io/stages/filters.returns.html"};
+
+CREATE_STATIC_STAGE(ReturnsFilter, s_info)
+
+std::string ReturnsFilter::getName() const
+{
+    return s_info.name;
+}
+
+void ReturnsFilter::addArgs(ProgramArgs& args)
+{
+    args.add("groups",
+             "Comma-separated list of return number groupings ('first', "
+             "'last', 'intermediate', or 'only')",
+             m_returnsString, {"last"});
+}
+
+void ReturnsFilter::prepared(PointTableRef table)
+{
+    const PointLayoutPtr layout(table.layout());
+    if (!layout->hasDim(Dimension::Id::ReturnNumber) ||
+        !layout->hasDim(Dimension::Id::NumberOfReturns))
+    {
+        log()->get(LogLevel::Warning)
+            << "Could not find ReturnNumber or "
+               "NumberOfReturns. Proceeding with all returns.\n";
+    }
+}
+
+PointViewSet ReturnsFilter::run(PointViewPtr inView)
+{
+    PointViewSet viewSet;
+    if (!inView->size())
+        return viewSet;
+
+    for (auto& r : m_returnsString)
+    {
+        Utils::trim(r);
+        if (r == "first")
+            m_outputTypes |= returnFirst;
+        else if (r == "intermediate")
+            m_outputTypes |= returnIntermediate;
+        else if (r == "last")
+            m_outputTypes |= returnLast;
+        else if (r == "only")
+            m_outputTypes |= returnOnly;
+        else
+            throwError("Invalid output type: '" + r + "'.");
+    }
+
+    PointViewPtr firstView = inView->makeNew();
+    PointViewPtr intermediateView = inView->makeNew();
+    PointViewPtr lastView = inView->makeNew();
+    PointViewPtr onlyView = inView->makeNew();
+
+    for (PointId idx = 0; idx < inView->size(); idx++)
+    {
+        PointRef p = inView->point(idx);
+        uint8_t rn = p.getFieldAs<uint8_t>(Dimension::Id::ReturnNumber);
+        uint8_t nr = p.getFieldAs<uint8_t>(Dimension::Id::NumberOfReturns);
+        if ((m_outputTypes & returnFirst) && (rn == 1) && (nr > 1))
+            firstView->appendPoint(*inView.get(), idx);
+        if ((m_outputTypes & returnIntermediate) && (rn > 1) && (rn < nr) &&
+            (nr > 2))
+            intermediateView->appendPoint(*inView.get(), idx);
+        if ((m_outputTypes & returnLast) && (rn == nr) && (nr > 1))
+            lastView->appendPoint(*inView.get(), idx);
+        if ((m_outputTypes & returnOnly) && (nr == 1))
+            onlyView->appendPoint(*inView.get(), idx);
+    }
+
+    if (firstView->size())
```

Comment:
It seems like if you've asked for certain sets, you should get them, regardless of whether they're empty or not.  I think this will omit the intermediate view, even if you've asked for it, as an example.

Reply:
I'm happy to make that change. I'm struggling to come up with a scenario in which an empty PointView would be useful. Would it be more useful to log a warning, at least acknowledging the request and explaining why it was not fulfilled?